src/nibbles.rs

```diff
@@ -635,123 +822,101 @@ impl Nibbles {
 
     /// Pushes a nibble to the end of the current nibbles without checking its validity.
     ///
-    /// Note that it is possible to create invalid [`Nibbles`] instances using this method. See
-    /// [the type docs](Self) for more details.
+    /// Note that only the low nibble of the byte is used. For example, for byte `0x12`, only the
+    /// nibble `0x2` is pushed.
     #[inline]
     pub fn push_unchecked(&mut self, nibble: u8) {
-        self.0.push(nibble);
+        let nibble_val = (nibble & 0x0F) as u64;
+        if nibble_val == 0 {
+            // Nothing to do, limb nibbles are already set to zero by default
+            self.length += 1;
+            return;
+        }
+
+        let bit_pos = (NIBBLES - self.length as usize - 1) * 4;
```

Comment:
replace all self.length manual field access with the method to make use of the assert_unchecked